src/main/java/com/launchdarkly/client/FeatureFlag.java

```diff
@@ -61,33 +61,32 @@ JsonElement getOffVariationValue() {
     return null;
   }
 
-  EvalResult evaluate(LDUser user, FeatureStore featureStore) {
+  EvalResult evaluate(LDUser user, FeatureStore featureStore) throws EvaluationError {
     if (user == null || user.getKey() == null) {
-      return null;
+      throw new EvaluationError("User or user key is null");
     }
     List<FeatureRequestEvent> prereqEvents = new ArrayList<>();
-    Set<String> visited = new HashSet<>();
-    JsonElement value = evaluate(user, featureStore, prereqEvents, visited);
+    JsonElement value = evaluate(user, featureStore, prereqEvents);
     return new EvalResult(value, prereqEvents);
   }
 
   // Returning either a JsonElement or null indicating prereq failure/error.
-  private JsonElement evaluate(LDUser user, FeatureStore featureStore, List<FeatureRequestEvent> events, Set<String> visited) {
+  private JsonElement evaluate(LDUser user, FeatureStore featureStore, List<FeatureRequestEvent> events) throws EvaluationError {
     boolean prereqOk = true;
-    visited.add(key);
     for (Prerequisite prereq : prerequisites) {
-      if (visited.contains(prereq.getKey())) {
-        logger.error("Prerequisite cycle detected when evaluating feature flag: " + key);
-        return null;
-      }
       FeatureFlag prereqFeatureFlag = featureStore.get(prereq.getKey());
       JsonElement prereqEvalResult = null;
       if (prereqFeatureFlag == null) {
         logger.error("Could not retrieve prerequisite flag: " + prereq.getKey() + " when evaluating: " + key);
         return null;
       } else if (prereqFeatureFlag.isOn()) {
-        prereqEvalResult = prereqFeatureFlag.evaluate(user, featureStore, events, visited);
-        if (prereqEvalResult == null || !prereqEvalResult.equals(prereqFeatureFlag.getVariation(prereq.getVariation()))) {
+        prereqEvalResult = prereqFeatureFlag.evaluate(user, featureStore, events);
+        try {
```

Comment:
If prerequisite evaluation fails for whatever reason, we return the off variation, not an error. So we don't bubble up prereq evaluation exceptions.